src/main/java/com/varukha/onlinebookstore/service/shoppingCart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,105 @@
+import com.varukha.onlinebookstore.repository.book.BookRepository;
+import com.varukha.onlinebookstore.repository.cartitem.CartItemRepository;
+import com.varukha.onlinebookstore.repository.shoppingcart.ShoppingCartRepository;
+import com.varukha.onlinebookstore.service.shoppingcart.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public void create(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto save(CreateCartItemRequestDto requestDto) {
+        CartItem existedCartItem = getCartItemByBookId(requestDto.getBookId());
+        if (existedCartItem == null || !existedCartItem.getBook().getId()
+                .equals(requestDto.getBookId())) {
+            return createNewCartItem(requestDto);
+        } else {
+            return updateCartItemQuantityWhenSave(existedCartItem, requestDto.getQuantity());
+        }
+    }
+
+    @Override
+    public CartItem getCartItemById(Long id) {
+        return cartItemRepository.findById(id).orElse(null);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCartByUserId(Long id) {
+        return shoppingCartRepository.findByUserId(getUserId())
+                .map(shoppingCartMapper::toDto)
+                .orElseThrow();
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto updateCartItemQuantity(Long id,
+                                              UpdateCartItemRequestDto cartItemRequestDto) {
+        CartItem cartItemById = getCartItemById(id);
+        int requestedQuantity = cartItemRequestDto.getQuantity();
+        cartItemById.setQuantity(requestedQuantity);
+        CartItem savedCartItem = cartItemRepository.save(cartItemById);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto updateCartItemQuantityWhenSave(CartItem existingCartItem, int quantity) {
+        existingCartItem.setQuantity(existingCartItem.getQuantity() + quantity);
+        CartItem savedCartItem = cartItemRepository.save(existingCartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
```

Comment:
Do you really need it?

Reply:
Yes, I do. This method is responsible to increment cart item quantity when we add the same book to shopping cart. I didn't create separate cart item and just increment existed one.

Reply:
You can redirect to update method if cart item with such book already exists)